File: packages/javascript/lib/core/session.ts
import { collectKeys, createAnalyzer, prepareGroups } from './analyzer.js';
import type {
	Chki18nEntry,
	Chki18nFileFormat,
	Chki18nInput,
	Chki18nIssue,
	Chki18nOptions,
	Chki18nResolvedOptions,
	Chki18nResult,
	Chki18nSourceFile,
	TranslationGroups,
	TranslationMap
} from '../_types/global.js';

export type Chki18nSession = {
	/** The options every check of this session runs with. */
	readonly options: Chki18nResolvedOptions;
	/** Every locale the session holds. */
	readonly locales: string[];
	/** Every group the session holds, in scan order. */
	readonly groups: string[];
	/** The files the translations were read from, when they came from disk. */
	readonly files: Chki18nSourceFile[];
	/** Layout the translations came from, or `null` for in-memory input. */
	readonly fileFormat: Chki18nFileFormat | null;
	/** Keys of a group, target language first. */
	keys: (group?: string) => string[];
	/**
	 * The flattened translations of a group, keyed by locale. The objects are
	 * the session's own: read them freely, but write through `set` / `remove`.
	 */
	translations: (group?: string) => { [locale: string]: TranslationMap };
	/** One value, or `undefined` when that locale does not define the key. */
	get: (locale: string, key: string, group?: string) => any;
	/** Write a value and report what that key now looks like. */
	set: (locale: string, key: string, value: string, group?: string) => Chki18nIssue[];
	/** Drop a key from one locale, or from every locale, and re-check it. */
	remove: (key: string, options?: { locale?: string; group?: string }) => Chki18nIssue[];
	/** Check a single key. Cross-key checks are not reported here. */
	checkKey: (key: string, group?: string) => Chki18nIssue[];
	/** Check everything the session holds. Reads no files. */
	analyze: () => Chki18nResult;
	/** Replace the translations, keeping the options and the analyzer. */
	reset: (input: Chki18nInput) => void;
};

/**
 * Hold a set of translations and check them repeatedly.
 *
 * Options are resolved once, the translations are flattened once, and every
 * later call works on what is already in memory. Use it when this module owns
 * the data; when your own application owns it, `createAnalyzer().checkEntry`
 * takes the values directly and keeps a single source of truth.
 */
export function createSession(input: Chki18nInput, options?: Chki18nOptions): Chki18nSession {
	// The session flattens once, up front, so the analyzer never has to.
	const analyzer = createAnalyzer({ ...options, flattened: true });
	const inputIsFlat = options?.flattened === true;

	let groups: TranslationGroups = {};
	let groupNames: string[] = [];
	let localeNames: string[] = [];
	let files: Chki18nSourceFile[] = [];
	let fileFormat: Chki18nFileFormat | null = null;
	let sourceIssues: Chki18nIssue[] = [];
	let unusedKeys: string[] = [];

	const reset = (next: Chki18nInput): void => {
		const loadIssues: Chki18nIssue[] = [];

		groups = prepareGroups(next ?? {}, { ...analyzer.options, flattened: inputIsFlat }, loadIssues);
		groupNames = Object.keys(groups);
		files = next?.files ?? [];
		fileFormat = next?.fileFormat ?? null;
		sourceIssues = [...(next?.issues ?? []), ...loadIssues];
		unusedKeys = next?.unusedKeys ?? [];

		const locales = new Set<string>();

		for (const group of groupNames) {
			for (const locale of Object.keys(groups[group])) {
				locales.add(locale);
			}
		}

		localeNames = [...locales];
	};

	/**
	 * Which group a call means. With one group there is nothing to decide; with
	 * several, an unnamed key is looked for where it actually lives, so callers
	 * only have to name a group when adding a key that does not exist yet.
	 */
	const resolveGroup = (key?: string, group?: string): string => {
		if (group !== undefined) {
			return group;
		}

		if (groupNames.length < 2 || key === undefined) {
			return groupNames[0] ?? '';
		}

		for (const name of groupNames) {
			for (const locale of Object.keys(groups[name])) {
				if (Object.hasOwn(groups[name][locale], key)) {
					return name;
				}
			}
		}

		return groupNames[0] ?? '';
	};

	const localesOf = (group: string): string[] => Object.keys(groups[group] ?? {});

	const checkKey = (key: string, group?: string): Chki18nIssue[] => {
		const name = resolveGroup(key, group);
		const locales = localesOf(name);
		const entry: Chki18nEntry = { key, values: {}, locales, group: name };

		for (const locale of locales) {
			if (Object.hasOwn(groups[name][locale], key)) {
				entry.values[locale] = groups[name][locale][key];
			}
		}

		return analyzer.checkEntry(entry);
	};

	reset(input ?? {});

	return {
		get options() {
			return analyzer.options;
		},
		get locales() {
			return localeNames;
		},
		get groups() {
			return groupNames;
		},
		get files() {
			return files;
		},
		get fileFormat() {
			return fileFormat;
		},
		keys: (group) => {
			const name = resolveGroup(undefined, group);
			const locales = localesOf(name);

			return collectKeys(
				locales.map((locale) => groups[name][locale]),
				locales.indexOf(analyzer.options.target)
			);
		},
		translations: (group) => groups[resolveGroup(undefined, group)] ?? {},
		get: (locale, key, group) => groups[resolveGroup(key, group)]?.[locale]?.[key],
		set: (locale, key, value, group) => {
			const name = resolveGroup(key, group);

			((groups[name] ??= {})[locale] ??= {})[key] = value;

			if (!localeNames.includes(locale)) {
				localeNames = [...localeNames, locale];
			}

			if (!groupNames.includes(name)) {
				groupNames = [...groupNames, name];
			}

			return checkKey(key, name);
		},
		remove: (key, removeOptions) => {
			const name = resolveGroup(key, removeOptions?.group);
			const locales = removeOptions?.locale ? [removeOptions.locale] : localesOf(name);

			for (const locale of locales) {
				delete groups[name]?.[locale]?.[key];
			}

			return checkKey(key, name);
		},
		checkKey,
		analyze: () =>
			analyzer.analyze({
				groups,
				files,
				issues: sourceIssues,
				unusedKeys,
				fileFormat: fileFormat ?? undefined
			}),
		reset
	};
}

File: packages/dart/lib/src/core/session.dart
/// Holding a set of translations and checking them repeatedly.
library;

import 'package:chki18n/src/constants.dart';
import 'package:chki18n/src/core/analyzer.dart';
import 'package:chki18n/src/types.dart';

/// A set of translations held in memory, checked as often as needed.
///
/// Options are resolved once, the translations are flattened once, and every
/// later call works on what is already in memory. Use it when this module owns
/// the data; when your own application owns it, [Chki18nAnalyzer.checkEntry]
/// takes the values directly and keeps a single source of truth.
class Chki18nSession {
  /// Creates a session over translations that are already in memory.
  ///
  /// [options] shapes every check the session runs. Set `flattened` on it when
  /// the maps handed in are already keyed by their dotted paths.
  Chki18nSession(Chki18nInput input, {Chki18nOptions? options})
    // The session flattens once, up front, so the analyzer never has to.
    : _analyzer = Chki18nAnalyzer(
        options: (options ?? const Chki18nOptions()).copyWith(flattened: true),
      ),
      _inputIsFlat = options?.flattened == true {
    reset(input);
  }

  final Chki18nAnalyzer _analyzer;
  final bool _inputIsFlat;

  TranslationGroups _groups = {};
  List<String> _groupNames = [];
  List<String> _localeNames = [];
  List<Chki18nSourceFile> _files = const [];
  Chki18nFileFormat? _fileFormat;
  List<Chki18nIssue> _sourceIssues = const [];
  List<String> _unusedKeys = const [];
  List<Chki18nKeyUsage> _undefinedKeys = const [];

  /// The options every check of this session runs with.
  Chki18nResolvedOptions get options => _analyzer.options;

  /// Every locale the session holds.
  List<String> get locales => List.unmodifiable(_localeNames);

  /// Every group the session holds, in scan order.
  List<String> get groups => List.unmodifiable(_groupNames);

  /// The files the translations were read from, when they came from disk.
  List<Chki18nSourceFile> get files => _files;

  /// Layout the translations came from, or `null` for in-memory input.
  Chki18nFileFormat? get fileFormat => _fileFormat;

  /// Replaces the translations, keeping the options and the analyzer.
  void reset(Chki18nInput input) {
    final loadIssues = <Chki18nIssue>[];

    _groups = prepareGroups(input, _analyzer.options, loadIssues, flattened: _inputIsFlat);
    _groupNames = _groups.keys.toList();
    _files = input.files ?? const [];
    _fileFormat = input.fileFormat;
    _sourceIssues = [...?input.issues, ...loadIssues];
    _unusedKeys = input.unusedKeys ?? const [];
    _undefinedKeys = input.undefinedKeys ?? const [];

    final seen = <String>{};

    for (final group in _groupNames) {
      seen.addAll(_groups[group]!.keys);
    }

    _localeNames = seen.toList();
  }

  /// Which group a call means.
  ///
  /// With one group there is nothing to decide; with several, an unnamed key is
  /// looked for where it actually lives, so callers only have to name a group
  /// when adding a key that does not exist yet.
  String _resolveGroup(String? key, String? group) {
    if (group != null) {
      return group;
    }

    if (_groupNames.length < 2 || key == null) {
      return _groupNames.isEmpty ? '' : _groupNames.first;
    }

    for (final name in _groupNames) {
      for (final locale in _groups[name]!.keys) {
        if (_groups[name]![locale]!.containsKey(key)) {
          return name;
        }
      }
    }

    return _groupNames.isEmpty ? '' : _groupNames.first;
  }

  List<String> _localesOf(String group) => _groups[group]?.keys.toList() ?? const [];

  /// Keys of a group, target language first.
  List<String> keys([String? group]) {
    final name = _resolveGroup(null, group);
    final locales = _localesOf(name);

    return collectKeys([
      for (final locale in locales) _groups[name]![locale]!,
    ], locales.indexOf(_analyzer.options.target));
  }

  /// The flattened translations of a group, keyed by locale.
  ///
  /// The maps are the session's own: read them freely, but write through [set]
  /// and [remove].
  Map<String, TranslationMap> translations([String? group]) =>
      _groups[_resolveGroup(null, group)] ?? <String, TranslationMap>{};

  /// One value, or `null` when that locale does not define the key.
  Object? get(String locale, String key, [String? group]) =>
      _groups[_resolveGroup(key, group)]?[locale]?[key];

  /// Writes a value and reports what that key now looks like.
  List<Chki18nIssue> set(String locale, String key, String value, [String? group]) {
    final name = _resolveGroup(key, group);

    ((_groups[name] ??= <String, TranslationMap>{})[locale] ??= <String, Object?>{})[key] = value;

    if (!_localeNames.contains(locale)) {
      _localeNames = [..._localeNames, locale];
    }

    if (!_groupNames.contains(name)) {
      _groupNames = [..._groupNames, name];
    }

    return checkKey(key, name);
  }

  /// Drops a key from one locale, or from every locale, and re-checks it.
  List<Chki18nIssue> remove(String key, {String? locale, String? group}) {
    final name = _resolveGroup(key, group);
    final locales = locale != null ? [locale] : _localesOf(name);

    for (final one in locales) {
      _groups[name]?[one]?.remove(key);
    }

    return checkKey(key, name);
  }

  /// Checks a single key. Cross-key checks are not reported here.
  List<Chki18nIssue> checkKey(String key, [String? group]) {
    final name = _resolveGroup(key, group);
    final locales = _localesOf(name);
    final values = <String, Object?>{};

    for (final locale in locales) {
      if (_groups[name]![locale]!.containsKey(key)) {
        values[locale] = _groups[name]![locale]![key];
      }
    }

    return _analyzer.checkEntry(
      Chki18nEntry(key: key, values: values, locales: locales, group: name),
    );
  }

  /// Checks everything the session holds. Reads no files.
  Chki18nResult analyze() => _analyzer.analyze(
    Chki18nInput(
      groups: _groups,
      files: _files,
      issues: _sourceIssues,
      unusedKeys: _unusedKeys,
      undefinedKeys: _undefinedKeys,
      fileFormat: _fileFormat,
    ),
  );
}

/// Holds a set of translations and checks them repeatedly.
Chki18nSession createSession(Chki18nInput input, {Chki18nOptions? options}) =>
    Chki18nSession(input, options: options);
